src/rust/ide/src/controller/project.rs

```diff
@@ -0,0 +1,244 @@
+// src/js/lib/project-manager/src/build.ts. See also https://github.com/enso-org/ide/issues/1359
+pub const ENGINE_VERSION_FOR_NEW_PROJECTS : &str = "0.2.11";
+
+/// The name of the module initially opened in the project view.
+///
+/// Currently, this name is hardcoded in the engine services and is populated for each project
+/// created using engine's Project Picker service.
+pub const INITIAL_MODULE_NAME:&str = "Main";
+
+/// Name of the main definition.
+///
+/// This is the definition whose graph will be opened on IDE start.
+pub const MAIN_DEFINITION_NAME:&str = "main";
+
+/// The code with definition of the default `main` method.
+pub fn default_main_method_code() -> String {
+    format!(r#"{} = "Hello, World!""#, MAIN_DEFINITION_NAME)
+}
+
+/// The default content of the newly created initial main module file.
+pub fn default_main_module_code() -> String {
+    default_main_method_code()
+}
+
+/// Method pointer that described the main method, i.e. the method that project view wants to open
+/// and which presence is currently required.
+pub fn main_method_ptr(project_name:impl Str, module_path:&model::module::Path) -> MethodPointer {
+    module_path.method_pointer(project_name,MAIN_DEFINITION_NAME)
+}
+
+
+// ==============
+// === Handle ===
+// ==============
+
+// === SetupResult ===
+
+/// The result of initial project setup, containing a handy controllers to be used in the initial
+/// view.
+#[derive(Clone,CloneRef,Debug)]
+pub struct SetupResult {
+    /// The Text Controller for Main module code to be displayed in Code Editor.
+    pub main_module_text:controller::Text,
+    /// The Graph Controller for main definition's graph, to be displayed in Graph Editor.
+    pub main_graph:controller::ExecutedGraph,
+}
+
+/// Project Controller Handle.
+///
+/// This controller supports IDE-related operations on a specific project.
+#[allow(missing_docs)]
+#[derive(Clone,CloneRef,Debug)]
+pub struct Project {
+    pub logger               : Logger,
+    pub model                : model::Project,
+    pub status_notifications : StatusNotificationPublisher,
+}
+
+impl Project {
+    /// Create a controller of given project.
+    pub fn new(model:model::Project, status_notifications:StatusNotificationPublisher) -> Self {
+        let logger = Logger::new("controller::Project");
+        Self {logger,model,status_notifications}
+    }
+
+    /// Do the initial setup of opened project.
+    ///
+    /// This function should be called always after opening a new project in IDE. It checks if main
+    /// module and main method are present in the project, and recreate them if they are missing.
+    /// It also sent proper status notifications and warnings related to the opened project (like
+    /// warning about unsupported engine version).
+    ///
+    /// Returns the controllers of module and graph which should be displayed in the view.
+    pub async fn initial_setup(&self) -> FallibleResult<SetupResult> {
```

Comment:
Perhaps that should be renamed just to `setup`? We don't have a `secondary_setup` after all.

Reply:
Will rename to initialize. (and rename SetupResult as well).